option to set recommended length for Essay questions #5606

tarun14110 opened the pull request:
Fixes #5606

![screenshot from 2016-06-15 10 40 25](https://cloud.githubusercontent.com/assets/12219700/16068693/a2ea4244-32e5-11e6-839f-8f8528397ad8.png)


![screenshot from 2016-06-15 10 39 24](https://cloud.githubusercontent.com/assets/12219700/16068714/ce369696-32e5-11e6-9432-a2fbeba2898a.png)

damithc commented:
Let us know when ready for review. Also, post screenshots.

tarun14110 commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -527,9 +539,9 @@ public void setQuestionDetails(FeedbackQuestionDetails questionDetails) {
         // This is due to legacy data in the data store before there are multiple question types
-        if (questionType == FeedbackQuestionType.TEXT) {
+        if (questionType == FeedbackQuestionType.TEXT && !isJsonValid(questionMetaData.getValue())) {
             return new FeedbackTextQuestionDetails(questionMetaData.getValue());
```
this is to handle previous Questions i.e that contain question as Text not as Json string.

tarun14110 commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,30 @@ public void sanitizeForSaving() {
-            questionMetaData = new Text(gson.toJson(questionDetails, getFeedbackQuestionDetailsClass()));
-        }
+        Gson gson = Utils.getTeammatesGson();
+        questionMetaData = new Text(gson.toJson(questionDetails, getFeedbackQuestionDetailsClass()));
```
previously for Text Questions we are simply saving question as Text, now saving that as Json string.

tarun14110 commented:
ready for review

JosephineKwa commented:
It would be good if you could left-align the "Recommended text length: ##" with the text area

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackTextQuestionDetails.java:
```diff
@@ -6,26 +6,34 @@
 
 public class FeedbackTextQuestionDetails extends FeedbackQuestionDetails {
     
+    public int recommendedLength;
```
Could this be made private instead?

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,10 @@
+                        name="recommendedlength" value="${recommendedlength}">
+        </div>
+    </div>
+    <br>
```
Perhaps this `<br>` could be removed so that the margin from the length input to the bottom of the purple box is equal to the margin from the top of the purple box to the textarea

damithc commented:
Recommended length input box seems to be too wide. It can be something like:
`Recommended length for the answer: [_______]`

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,10 @@
+    <br>
+    <div class="row">
+        <div class="col-sm-4" data-toggle="tooltip" data-placement="top" title="recommended length ">Recommended Length:
+            <input disabled type="number" class="form-control" id="recommendedlength"
```
Do remove this id if possible

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,10 @@
+<div>
+    <br>
+    <div class="row">
+        <div class="col-sm-4" data-toggle="tooltip" data-placement="top" title="recommended length ">Recommended Length:
```
Do indicate somewhere that the length refers to the number of chars as well :)

JosephineKwa replied:
Also I feel that the tooltip wouldn't be necessary as it is a duplicate of the text

damithc replied:
The tool tip can indicate that the recommended length is shown to the respondent but not enforced.

damithc commented:
Is this better?
`Recommended length for the answer: [_______] letters`
On a related note, should we use words instead? I think text length is more commonly measured in words.

chowyb commented:
Agree on using words for text length instead. Length based on character count seems to be more of a technical perspective (eg. checking if a given input can fit within the maximum length of a string in the database) as opposed to the user's perspective (checking if a response meets the minimum/maximum word count, although as noted earlier the length won't be enforced here).

On the other hand, keeping track of word count may be slightly harder once rich text comes into play.

JosephineKwa commented:
Will resume the review later/tomorrow

tarun14110 commented:
agree, generally words count are used to represent text length limit, words count would give better idea about the length of the response. Recommending length in terms of char counts is much more difficult.

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,30 @@ public void sanitizeForSaving() {
         this.creatorEmail = Sanitizer.sanitizeEmail(creatorEmail);
     }
+    
+    public boolean isJsonValid(String test) {
```
Perhaps rename `test` to `json` or `jsonString`?

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,30 @@ public void sanitizeForSaving() {
+            new JSONObject(test);
+        } catch (JSONException ex) {
+            // edited, to include @Arthur's comment
+            // e.g. in case JSONArray is valid as well...
```
Could remove the top comment :p (line 515), rephrase the bottom comment (line 516) to be `Checks whether JSONArray is valid` and move it to be just above `new JSONArray(test);`

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackTextQuestionDetails.java:
```diff
@@ -6,26 +6,35 @@
+        String recommendedLengthString = HttpRequestHelper.getValueFromParamMap(requestParameters,
+                Const.ParamsNames.FEEDBACK_QUESTION_TEXT_RECOMMENDEDLENGTH);
+
+        this.recommendedLength = recommendedLengthString == null || recommendedLengthString.isEmpty() ? 0
```
As the `this` keyword is not necessary here, I would prefer to remove it (and at other parts where it is not necessary). This is also to maintain consistency as some `recommendedLength` are prefixed with `this` but some aren't.

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -975,7 +975,17 @@ function getWarningMessage() {
+    } else {
+        $('#' + charCountId).css('color', 'gray');
+    }
+	}
```
Align the curly brace with `function`

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -527,9 +539,9 @@ public void setQuestionDetails(FeedbackQuestionDetails questionDetails) {
+        // For old Text questions, the questionText simply contains the question, not a JSON
         // This is due to legacy data in the data store before there are multiple question types
-        if (questionType == FeedbackQuestionType.TEXT) {
+        if (questionType == FeedbackQuestionType.TEXT && !isJsonValid(questionMetaData.getValue())) {
```
Maybe extract `questionMetaData.getValue()` into a variable as it is used thrice in this method

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,30 @@ public void sanitizeForSaving() {
+    public boolean isJsonValid(String test) {
+        try {
+            new JSONObject(test);
+        } catch (JSONException ex) {
```
Rename both exceptions as `e`

JosephineKwa commented:
Would you prefer to change the response text length (added in #5556) to count words in this PR as well or in a separate PR? If you will be doing it in a separate PR, I will put this on hold until that has been merged :)

tarun14110 commented:
@JosephineKwa will do within this PR. I had updated the suggested changes.

tarun14110 commented:
![screenshot from 2016-06-17 04 34 38](https://cloud.githubusercontent.com/assets/12219700/16136049/0e1636c6-3445-11e6-9aca-b497c16b788b.png)


![screenshot from 2016-06-17 04 36 05](https://cloud.githubusercontent.com/assets/12219700/16136050/0e1b2ec4-3445-11e6-9ee0-0b39f812a1af.png)

damithc commented:
We need to indicate that the setting is optional.
Prefix with `[Optional] ...`?

tarun14110 commented:
@damithc better ?
![screenshot from 2016-06-17 13 25 55](https://cloud.githubusercontent.com/assets/12219700/16144271/6ff9f5c2-348f-11e6-9f7d-4b7ff96582a7.png)

damithc commented:
Yes. the text box seems too wide though. I don't think we need more than 5 digits' width. Probably need an `(i)` icon linked to an explanation in the `instructorHelp` page? Otherwise some users will not know how this field affects the behavior.

tarun14110 commented:
reduced recommended textbox width
![screenshot from 2016-06-17 19 45 09](https://cloud.githubusercontent.com/assets/12219700/16153452/07d62fd0-34c4-11e6-9090-4f5029d4e927.png)

JosephineKwa commented:
@tarun14110 , just checking, is this PR ready for review?

tarun14110 commented:
@JosephineKwa I had missed the (i) icon linked to an explanation in the instructorHelp page. will update and get back to you.

tarun14110 commented:
@damithc do we really need  icon linked to an explanation ? Tooltip is giving enough information and its much clear that it is optional and will not be enforced on the student ?

damithc commented:
In that case, can we apply the tooltip to the term `recommended length` only? Also, underline the term with grey dotted line to make it more obvious. You can follow this example from the results page:
![image](https://cloud.githubusercontent.com/assets/1673303/16251629/37dfc670-385a-11e6-9f7c-a3a005edfc1a.png)

tarun14110 commented:
better ?
![screenshot from 2016-06-22 14 21 56](https://cloud.githubusercontent.com/assets/12219700/16260358/b6780d56-3884-11e6-9a7d-9096cd484439.png)

damithc commented:
UI looks fine but I don't think we should add the style directly like that. @YongJieYongJie what do you suggest?

tarun14110 commented:
@JosephineKwa ready for review again.

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -134,6 +136,10 @@ Apply to thead to make headers not bold.
+.float-left {
+    float: left;
+}
+
```
Unless absolutely necessary, I think we shouldn't create classes that are actually just the rules. It defeats the purpose of having a class, and still violates the separation of styling concerns from HTML.

Perhaps we can create a class on the surrounding div (e.g. `".question-recommended-length"`) and apply styles using the class and child selectors (since the related DOM structure here is rather simple).

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -227,10 +233,22 @@ Apply to thead to make headers not bold.
+.padding-top-7 {
+    padding-top: 7px;
+}
+
```
Again, I think we should not create classes that directly maps to a CSS rule.

I understand that there are similar classes in our existing css, but (I suspect) they actually takes two other forms and are thus not merely direct mapping to CSS rules:

1. They may be edge/special values like 0, 100%, 75/25% etc., or
2. They actually have semantic meaning (e.g., "margin after question panel" or "question form padding"), and are not for one-off use.

(1 is actually a subset of 2, for example, the `.margin-0` below actually has the semantic meaning of "no margin")

In any case, perhaps an alternative approach is to use the `.question-recommended-length` class as suggested above.

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -303,6 +321,10 @@ Apply to thead to make headers not bold.
+.width-80-px {
+    width: 80px;
+}
+
```
Same as above

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -63,7 +63,9 @@ a#editUncommonSettingsButton:hover{
-
+.border-bottom-dotted {
+    border-bottom: 1px dotted;
+}
```
Since we are going for an underlined text, is it possible to use `text-decoration: underline;` instead?

tarun14110 replied:
we need dotted underline. And `text-decoration-style` is not supported in most of the browsers.

YongJieYongJie commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,17 @@
+            <div class="col-xs-3 width-80-px padding-0 margin-right-10">
+                <input disabled type="number" class="form-control" name="recommendedlength" value="${recommendedlength}">
+            </div>
+            <label class="padding-top-7 text-color-gray font-weight-normal">words<label>
```
Since the whole line of text is conceptually one "thing", is it possible to not break it into so many labels. Perhaps something like
```
<div class="question-recommended-length">
    ... <span class="underline-dotted">Recommended length</span> ... <input> ...
</div>
```

YongJieYongJie commented:
@damithc @tarun14110 Added comments.

Generally, since the rules are localized to a specific group of elements, we should create a class for the surrounding div and apply styles relative to that surrounding div.

tarun14110 commented:
@YongJieYongJie please have a look over recent changes. are they fine ?

tarun14110 commented:
@JosephineKwa ready for review

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -259,6 +259,20 @@ Apply to thead to make headers not bold.
+
+.question-recommended-length > span {
+    margin-left: 7px;
+    border-bottom: 1px dotted;
```
Does `text-decoration: underline` not work?

tarun14110 replied:
we need dotted underline.

YongJieYongJie replied:
Yup, I think it's better to use that as its semantics better fit what we are trying to achieve

YongJieYongJie replied:
@tarun14110 Please do change this to use text-decoration for the underline

YongJieYongJie commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -259,6 +259,20 @@ Apply to thead to make headers not bold.
     padding-top: 0;
 }
 
+.question-recommended-length {
```
Is there any better place to put these rules? Currently it seems to be in between some random rules

tarun14110 replied:
I was trying to follow the alphabetical order. And it seems there are not such classes defined in this  file. Otherwise I can shift these classes to the starting or end if that makes sense ?

YongJieYongJie replied:
I see I see. The `teammatesCommon.css` can indeed by re-organized. For now, maybe put these somewhere near `.session-name-table-width` or other component-related rules

YongJieYongJie commented on src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html:
```diff
@@ -1,9 +1,16 @@
+    <div class="pull-right">
+        <label class="text-color-gray font-weight-normal">Response length: </label>
+        <label class="text-color-gray font-weight-normal" id="responseLength-${questionIndex}-${responseIndex}">0</label>
+        <label class="text-color-gray font-weight-normal"> words</label>
```
I think can avoid having too many `<label>` tags here too.

YongJieYongJie commented:
@tarun14110 Added comments relating to CSS and the HTML template

tarun14110 commented:
@JosephineKwa ready for review again

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,16 @@
+        <div class="col-xs-11 question-recommended-length">
+            [Optional]
+            <span data-toggle="tooltip" data-placement="top"
+                title="the recommended length is shown to the respondent but not enforced">
```
Capitalise the first `the`

YongJieYongJie commented:
@tarun14110 do post a screenshot (I think the bottom margin is too big, and there should be a space after `for the answer:` and before the input field)

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,16 @@
+<div>
+    <br>
+    <div class="row">
+        <div class="col-xs-11 question-recommended-length">
```
Could this be 12 instead of 11?

JosephineKwa commented on src/main/resources/feedbackQuestionTextEditFormTemplate.html:
```diff
@@ -0,0 +1,16 @@
+            <span data-toggle="tooltip" data-placement="top"
+                title="the recommended length is shown to the respondent but not enforced">
+                Recommended length</span>
+            for the answer:
```
Maybe change this to `for the response` so that the wording is consistent with the wording of the response length count shown in the submission edit page.

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +962,22 @@ function getWarningMessage() {
-function updateTextQuestionCharCount(textAreaId, charCountId) {
-    var cs = $('#' + textAreaId).val().length;
-    $('#' + charCountId).text(cs);
+function updateTextQuestionCharCount(textAreaId, charCountId, recommendedLength) {
```
Do update the names of the method and variables to reflect that words are being counted rather than characters

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +962,22 @@ function getWarningMessage() {
-    var cs = $('#' + textAreaId).val().length;
-    $('#' + charCountId).text(cs);
+function updateTextQuestionCharCount(textAreaId, charCountId, recommendedLength) {
+    var cs = $('#' + textAreaId).val().trim().split(/\s+/).length;
```
As `$('#' + textAreaId)` and `$('#' + charCountId)` are used multiple times, it would be good to extract them into variables

JosephineKwa commented:
Currently a sequence of symbols surrounded with whitespace is also counted as a word, could you change the method such that only a sequence containing alphanumeric characters is counted as a word? It can still contain symbols, i.e. `Alice's`, but it has to contain at least one alphanumeric character.

tarun14110 commented:
@YongJieYongJie I updated the bottom margin. @JosephineKwa ready for review again.
![screenshot from 2016-06-27 13 57 26](https://cloud.githubusercontent.com/assets/12219700/16408621/409bcade-3d37-11e6-8f36-eade9c5b9ec4.png)

tarun14110 commented:
@YongJieYongJie I updated the bottom margin. @JosephineKwa ready for review again.updated as suggested by you.
![screenshot from 2016-06-27 13 57 26](https://cloud.githubusercontent.com/assets/12219700/16408621/409bcade-3d37-11e6-8f36-eade9c5b9ec4.png)

JosephineKwa commented on src/main/resources/feedbackQuestionTextSubmissionFormTemplate.html:
```diff
@@ -1,9 +1,17 @@
-    <label class="text-color-gray font-weight-normal" id="responseLength-${questionIndex}-${responseIndex}">0</label>
-    <label class="text-color-gray font-weight-normal"> characters</label>
-</div>
+    data-lengthTextId="responseLength-${questionIndex}-${responseIndex}" data-recommendedText="${recommendedLength}">${existingResponse}</textarea>
```
I missed this out earlier, it would be preferred if you used hyphens to delimit the data attribute. i.e. `data-length-text-id` and `data-recommended-text`

JosephineKwa commented:
Should be good to merge after addressing the comment

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -16,11 +16,11 @@ $(document).ready(function() {
      * Handles Keyup and Keydown on Text question to display response length
      */
     $('textarea[id^="responsetext-"]').keyup(function() {
```
Not entirely related to this PR, but perhaps you could use `.on('input', function() {...})` instead of `.keyup` and `.keydown`. That way, it also factors in the cases where the user cuts or pastes into the `textarea` with a mouse. Create a separate issue for it maybe?

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,22 +962,28 @@ function getWarningMessage() {
+function updateTextQuestionWordsCount(textAreaId, wordsCountId, recommendedLength) {
+	
+    var response = $('#' + textAreaId).val();
+    var wordsCountElement = $('#' + wordsCountId);
```
prefix this variable with `$` to show that it is a jQuery object

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,22 +962,28 @@ function getWarningMessage() {
-            && $('#' + charCountId).text() < upperLimit) {
-        $('#' + charCountId).css('color', 'green');
+    if (wordsCountElement.text() > lowerLimit
+            && wordsCountElement.text() < upperLimit) {
```
Extract `wordsCountElement.text()` into a variable too

tarun14110 commented:
@JosephineKwa ready for review again.

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -15,12 +15,12 @@ $(document).ready(function() {
 
-    $('textarea[id^="responsetext-"]').keydown(function() {
-        updateTextQuestionCharCount(this.id, $(this).data('lengthtextid'));
+    $('textarea[id^="responsetext-"]').on('input', function() {
```
No need for this as it is a duplicate of the above

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +962,29 @@ function getWarningMessage() {
-    $('#' + charCountId).text(cs);
+function updateTextQuestionWordsCount(textAreaId, wordsCountId, recommendedLength) {
+	
+    var $response = $('#' + textAreaId).val();
```
Not necessary for `$response` to have `$` as it is not a jQuery object

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +962,29 @@ function getWarningMessage() {
+	
+    var $response = $('#' + textAreaId).val();
+    var $wordsCountElement = $('#' + wordsCountId);
+    var $wordsCountElementText = $('#' + wordsCountId).text();
```
Similarly, not necessary for `$wordscountElementText` to have `$` as it is not a jQuery object

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +962,29 @@ function getWarningMessage() {
+    var upperLimit = recommendedLength + recommendedLength * 0.1;
+    var lowerLimit = recommendedLength - recommendedLength * 0.1;
+
+    if ($wordsCountElementText > lowerLimit
```
Should be ok to put the below ` && $wordsCountElementText < upperLimit) {` on the same line as the above

tarun14110 commented:
@JosephineKwa have a look

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +958,28 @@ function getWarningMessage() {
+    if (response.length === 0) {
+        $wordsCountElement.text('0');
+    } else {
+        $wordsCountElement.text(wordsCount);
```
Would it be ok to simply do `$wordsCountElement.text(wordsCount);` without checking if `response.length === 0`?

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -962,7 +958,28 @@ function getWarningMessage() {
+    var upperLimit = recommendedLength + recommendedLength * 0.1;
+    var lowerLimit = recommendedLength - recommendedLength * 0.1;
+
+    if (wordsCountElementText > lowerLimit && wordsCountElementText < upperLimit) {
```
I realised we could just use `wordsCount` here too, instead of `wordsCountElementText` :disappointed:

tarun14110 replied:
sorry for that. removed `wordsCountElementText`.

JosephineKwa commented:
It looks good to merge :)

damithc commented:
@kanghj what's your input on the json handling? Looks a bit hackish to me. This is in fact a general problem of json schema changing over time. We should have a standard way to deal with it.

JosephineKwa commented:
Sorry for overlooking this, would existing responses have to be migrated, and would existing text questions in json files also have to be edited?

damithc commented:
I didn't realize one question type was saved in a different format. Anyway, migration is a last resort. The ideal workflow is to 
1. deploy a version that can work with both formats
2. After a while, migrate old data into new format
3. Remove code used to handle old format.

Coming back to this PR, it's fine if the code can handle both types. I'm wondering if there is a better way to handle the two branches.

damithc commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,29 @@ public void sanitizeForSaving() {
+            }
+        }
+        return true;
+    }
```
Any better way to do this? Using two exceptions seems hackish.

kanghj commented:
The json handling does not look good to me too, but it should be fine if we can migrate the old Text question data soon. 

If we are not able to do that soon, then a possible alternative to checking if the question details is json is to temporarily replace TEXT questions with a new question type (e.g. `ESSAY`) in the meantime.  This will use the usual json format that we have. This will keep the code in `FeedbackQuestionAttributes` untouched. Eventually, we can migrate all the TEXT questions and completely remove that difficult branch handling the legacy format.

damithc commented:
Adding a new type may be more work than what we have here? I'm fine with the current code too. It's just that I thought there may be a better way to check if the string is in the correct json format instead of using two catch blocks. Perhaps we should treat it as valid json (new format) in the `try` branch and handle the old format data in the `catch` branch?
Something like:
```java
try{
    //convert to json object
} catch (Exception){
   //if TEXT type, check if data is in old format
}
```

tarun14110 commented:
Is it better ?
```
public boolean isJsonValid(String jsonString) {
    try{
        JsonParser parser = new JsonParser();
        parser.parse(jsonString);
    } 
    catch(JsonSyntaxException e){
        return false;
    }
    return true;
}
```

damithc commented:
yes it looks neater than the previous one. 
`catch` should be in the same line as `}`

tarun14110 commented:
@JosephineKwa ready again.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -510,14 +509,10 @@ public void sanitizeForSaving() {
-                return false;
-            }
+            JsonParser parser = new JsonParser();
+            parser.parse(jsonString);
```
@tarun14110, this can't be right. Gson's JsonParser is very lenient so even `new JsonParser().parse("123");` does not throw an exception.

damithc replied:
On a related note, we need unit tests to ensure the code works for both old format and new format. i.e. There should be a test failing if this method is broken.

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -504,21 +507,24 @@ public void sanitizeForSaving() {
     }
-
+    
+    public boolean isValidJsonObject(String jsonString) {
```
Maybe rename this to `isValidJsonString` since the input is a `String`. Also, could make this method private.

JosephineKwa commented on src/test/java/teammates/test/cases/common/FeedbackQuestionAttributesTest.java:
```diff
@@ -170,6 +174,45 @@ public void testValidate() {
+        fq.showRecipientNameTo = new ArrayList<FeedbackParticipantType>();
+        fq.showRecipientNameTo.add(FeedbackParticipantType.RECEIVER);
+        fq.showResponsesTo = new ArrayList<FeedbackParticipantType>();
+        fq.showResponsesTo.add(FeedbackParticipantType.RECEIVER);
```
Instead of assigning these attributes manually, perhaps add a sample text question to `typicalDataBundle.json` which uses the new `questionMetaData` format and use that question to perform the assertion

JosephineKwa commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -539,6 +539,20 @@ Apply to thead to make headers not bold.
+.question-recommended-length > input {
+    width: 80px;
+    display: inline;
+    }
```
Remove the spaces before the `}`

JosephineKwa commented on src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java:
```diff
@@ -527,13 +533,14 @@ public void setQuestionDetails(FeedbackQuestionDetails questionDetails) {
-        // For Text questions, the questionText simply contains the question, not a JSON
+        final String questionMetaDataValue = questionMetaData.getValue();
+        // For old Text questions, the questionText simply contains the question, not a JSON
         // This is due to legacy data in the data store before there are multiple question types
```
Could remove this line

tarun14110 commented:
@JosephineKwa Can you please help me with the failed travis tests ?

JosephineKwa commented:
Hey @tarun14110 , the findBugs report when running  `./gradlew statisAnalysis` tells me that there is a violation at line 180 of FeedbackQuestionAttributesTest.java where there is a "Dead store to local variable". Instead of doing
```
FeedbackQuestionAttributes fq = new FeedbackQuestionAttributes();
fq = typicalBundle.feedbackQuestions.get("qn5InSession1InCourse1");
```
we should do 
```
FeedbackQuestionAttributes fq = typicalBundle.feedbackQuestions.get("qn5InSession1InCourse1");
```

JosephineKwa commented:
@tarun14110 you will need to resolve the conflicts so that Travis will run the build

tarun14110 commented:
@JosephineKwa ready again.

JosephineKwa commented on src/test/resources/data/typicalDataBundle.json:
```diff
@@ -875,6 +875,30 @@
+            "questionType":"TEXT",
+            "giverType":"SELF",
+            "recipientType":"NONE",
+            "numberOfEntitiesToGiveFeedbackTo":1,
```
This should be -100 instead of 1

JosephineKwa commented:
Minor issue to address

kanghj commented:
@tarun14110, remember to ping when ready for review.

tarun14110 commented:
@JosephineKwa ready again. Have a look.

JosephineKwa commented:
Looks good to merge! :)

tarun14110 commented:
@JosephineKwa resolved.

JosephineKwa commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -13,16 +13,12 @@ function isPreview() {
-
+    
     /**
      * Triggering keyup event for all text question type textfields, to call
```
Will need you to change `$('textarea[id^="responsetext-"]').keyup();` to `$('textarea[id^="responsetext-"]').trigger('input');`, do edit the comment to `input` event instead of `keyup` event also.

tarun14110 commented:
@JosephineKwa ready for review.

JosephineKwa commented:
@tarun14110 could you update the branch with the latest master? So that I will be able to merge this PR :)

tarun14110 commented:
@JosephineKwa updated.
